### ISSUE TITLE
Filesystem-based dynamic TCP forwarding.

### ISSUE THREAD

**Issue by askquestions** [envoyproxy/envoy#17342]
I could not make dynamic TCP forwarding to work. Envoyproxy failed in forwarding TCP packets. I post 3 configuration files: envoy.yaml, lds.yaml, and cds.yaml. It is hard to find helpful example configuration on handling TCP flows. Can anyone provide some helps. Thank you.

envoy.yaml:

```
node:
  id: id_1
  cluster: test

dynamic_resources:
  cds_config:
    path: /var/lib/envoy/cds.yaml
  lds_config:
    path: /var/lib/envoy/lds.yaml

admin:
  address:
    socket_address:
      address: 0.0.0.0
      port_value: 19000
```

lds.yaml:

```
resources:
- "@type": type.googleapis.com/envoy.config.listener.v3.Listener
  name: listener_0
  address:
    socket_address:
      address: 0.0.0.0
      port_value: 10000
  filter_chains:
  - filters:
     - name: envoy.http_connection_manager
       typed_config:
         "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
         stat_prefix: ingress_http
         http_filters:
           - name: envoy.router
             route_config:
               name: local_route
               virtual_hosts:
                  - name: local_service
                       domains:
                           - "*"
                       routes:
                          - match:
                               prefix: "/"
                               route:
                                  cluster: example_proxy_cluster
      - name: envoy.filters.network.tcp_proxy
           typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.network.tcp_proxy.v3.TcpProxy
              stat_prefix: tcp
              cluster: tcp_proxy_cluster
```

cds.yaml:

```
resources:
   - "@type": type.googleapis.com/envoy.config.cluster.v3.Cluster
      name: http_proxy_cluster
      type: STRICT_DNS
      load_assignment:
         cluster_name: http_proxy_cluster
         endpoints:
             - lb_endpoints:
                  - endpoint:
                       address:
                           socket_address:
                               address: service1
                               port_value: 8080

  - "@type": type.googleapis.com/envoy.config.cluster.v3.Cluster
     name: tcp_proxy_cluster
     type: STRICT_DNS
     load_assignment:
        cluster_name: tcp_proxy_cluster
        endpoints:
            - lb_endpoints:
                 - endpoint:
                       address:
                           socket_address:
                               address: service2
                               port_value: 10000
```

**Comment by lambdai** [comment]
1. Do you need to update yaml on the filesystem? If not I suggest you use single envoy.yaml

2. more log on the error will be helpful. 
    lds.yaml must be wrong: the two filters (tcpproxy and httpconnmanager doesn't indent. If the two filters belong to different filter chains, you need filter chain match to isolate the two.

**Comment by askquestions** [comment]
Thank you for the response. The configuration files (*.yaml) are actually correct. The problem was caused by the left-over garbages in the previously run containers. After we run docker-compose with options of deleting orphaned containers and other garbages, we have made the configuration to functioning.

Once again, our .yaml config files as shown in my post are correct.

**Comment by lambdai** [comment]
@askquestions I am not too sure. The indent might be lost in the paste.
It's clear that `local_service` is within the `virtualhost`
```
               virtual_hosts:
          - name: local_service
            domains:
```

Also you might want to ask what traffic would hit tcp proxy and what would hit http_conn_manager.

**Comment by askquestions** [comment]
> @askquestions I am not too sure. The indent might be lost in the paste.
> It's clear that `local_service` is within the `virtualhost`
> 
> ```
>                virtual_hosts:
>           - name: local_service
>             domains:
> ```
> 
> Also you might want to ask what traffic would hit tcp proxy and what would hit http_conn_manager.

Thank you for your comment. I have edited lds.yaml in my initial post. Do you feel that I made the correct indentation in the edited lds.file?

I am in the learning process of using envoy and consul. In my current project, I only need to use tcp filter. However, I want to test the configuration which can handle mixed types of traffics. For example, I wanted to test whether hybrid HTTP traffic and raw TCP traffic can be separately routed to different upstream servers even when both types of traffics are fed to a common listener port of an ingress envoy proxy.

In the meantime, I want to move forward to work on control plane based dynamic config on envoy. Do you have experience on the integration of envoy and consul? Right now, I am testing this repo: https://github.com/tak2siva/Envoy-Pilot

I made a reuse of lds.yaml as shown in my initial post, and I only modified the tcp part in cds.yaml for hopefully fetching the endpoint IP and port number from the xDS server. I have not made a success in performing this cp-based config and have some questions. If I can get some helps from you, then I can post the current content of cds.yaml using gRPC calls to xDS server.

Thank you very much.

**Comment by lambdai** [comment]
Great! The lds.yaml looks better. Now we can move to the next topic

>> when both types of traffics are fed to a common listener port of an ingress envoy proxy.

That make sense for ingress. However, you need something distinguish the traffic, http or tcp, ahead of the filter chain. Neither http_conn_manager nor TcpProxy disable itself when the traffic is not expected.

What you need a FilterChainMatch for each filter chain: If the traffic has these attribributes, create a filter chain that contains HCM to server the tcp connection; otherwise, instantiate a filter chain contains TcpProxy to serve.

The attributes can be populated by socket itself, e.g. destination ip, port,
Or the [listener filters](https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/listener_filters/listener_filters)

An example can be found [here](https://www.envoyproxy.io/docs/envoy/latest/intro/life_of_a_request.html?highlight=filter%20chain%20match#configuration).

**Comment by askquestions** [comment]
Sorry to get back to you with a delay. I have tried to add filter_chain_match into lds.yaml and to specify 2 clusters in cds.yaml. However, I did not succeed in making a working configuration. I post the contents of lds.yaml and cds.yaml.
lds.yaml
```
resources:
- "@type": type.googleapis.com/envoy.config.listener.v3.Listener
  name: listener_0
  address:
    socket_address:
      protocol: TCP
      address: 0.0.0.0
      port_value: 10000
#  listener_filters:
#  - name: "envoy.filters.listener.tls_inspector"
  filter_chains:
  - filter_chain_match:
# #     server_names: ["fakeserver_cluster"]
    filters:
      name: envoy.http_connection_manager
      typed_config:
        "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
        stat_prefix: ingress_http
        http_filters:
        - name: envoy.router
        route_config:
          name: local_route
          virtual_hosts:
          - name: local_service
            domains:
            - "*"
            routes:
            - match:
                prefix: "/"
              route:
                cluster: fakeserver_cluster
  - filter_chain_match:
      server_names: ["onos_one"]
    filters:
      name: envoy.filters.network.tcp_proxy
      typed_config:
        "@type": type.googleapis.com/envoy.extensions.filters.network.tcp_proxy.v3.TcpProxy
        stat_prefix: ingress_tcp
        cluster: onos_cluster
```
cds.yaml
```
resources:
- "@type": type.googleapis.com/envoy.config.cluster.v3.Cluster
  name: fakeserver_cluster
  type: STRICT_DNS
  load_assignment:
    cluster_name: fakeserver_cluster
    endpoints:
    - lb_endpoints:
      - endpoint:
          address:
            socket_address:
              address: fakeserver
              port_value: 8080
  name: onos_cluster
  type: STRICT_DNS
  load_assignment:
    cluster_name: onos_cluster
    endpoints:
    - lb_endpoints:
      - endpoint:
          address:
            socket_address:
              address: onos_one
              port_value: 6633

```

One comment is that the above configuration worked when only specifying one filter in lds.yaml and one cluster in cds.yaml. For example, a valid pair of lds.yaml and cds.yaml are as follows.
lds.yaml
```
resources:
- "@type": type.googleapis.com/envoy.config.listener.v3.Listener
  name: listener_0
  address:
    socket_address:
      protocol: TCP
      address: 0.0.0.0
      port_value: 10000
#  listener_filters:
#  - name: "envoy.filters.listener.tls_inspector"
  filter_chains:
  - filter_chain_match:
      server_names: ["onos_one"]
    filters:
      name: envoy.filters.network.tcp_proxy
      typed_config:
        "@type": type.googleapis.com/envoy.extensions.filters.network.tcp_proxy.v3.TcpProxy
        stat_prefix: ingress_tcp
        cluster: onos_cluster
```
cds.yaml
```
resources:
- "@type": type.googleapis.com/envoy.config.cluster.v3.Cluster
  name: onos_cluster
  type: STRICT_DNS
  load_assignment:
    cluster_name: onos_cluster
    endpoints:
    - lb_endpoints:
      - endpoint:
          address:
            socket_address:
              address: onos_one
              port_value: 6633
```
Meanwhile, another valid combination is:
lds.yaml
```
resources:
- "@type": type.googleapis.com/envoy.config.listener.v3.Listener
  name: listener_0
  address:
    socket_address:
      protocol: TCP
      address: 0.0.0.0
      port_value: 10000
#  listener_filters:
#  - name: "envoy.filters.listener.tls_inspector"
  filter_chains:
  - filter_chain_match:
# #     server_names: ["fakeserver_cluster"]
    filters:
      name: envoy.http_connection_manager
      typed_config:
        "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
        stat_prefix: ingress_http
        http_filters:
        - name: envoy.router
        route_config:
          name: local_route
          virtual_hosts:
          - name: local_service
            domains:
            - "*"
            routes:
            - match:
                prefix: "/"
              route:
                cluster: fakeserver_cluster
```
cds.yaml
```
resources:
- "@type": type.googleapis.com/envoy.config.cluster.v3.Cluster
  name: fakeserver_cluster
  type: STRICT_DNS
  load_assignment:
    cluster_name: fakeserver_cluster
    endpoints:
    - lb_endpoints:
      - endpoint:
          address:
            socket_address:
              address: fakeserver
              port_value: 8080
```
Basically, a working configuration can be made when a single filter specified in (lds.yaml) is used with a single cluster (specified in cds.yaml).

Did I made any mistake in the configurations? Do you you have any suggestion? Thanks.
